gbpservice/neutron/nsf/core/main.py

```diff
@@ -0,0 +1,705 @@
+    return 'class' in str(type(obj))
+
+
+def name(obj):
+    if callable(obj):
+        return "{0}.{1}.{2}".format(
+            type(obj.im_self).__module__,
+            type(obj.im_self).__name__,
+            obj.__name__)
+    elif is_class(obj):
+        return "{0}.{1}".format(
+            type(obj).__module__,
+            type(obj).__name__)
+    else:
+        return obj.__name__
+
+
+def identify(obj):
+    try:
+        return "(%s)" % (name(obj))
+    except:
+        return ""
+
+
+class AgentReportState(object):
+
+    def __init__(self, agent_state):
+        self._context = context.get_admin_context_without_session()
+        self._agent_state = agent_state
+        self._report_topic = agent_state['plugin_topic']
+        self._report_interval = agent_state['report_interval']
+        self._state_rpc = agent_rpc.PluginReportStateAPI(
+            self._report_topic)
+
+    def report_state(self):
+        try:
+            LOG.debug(_("Reporting state ", self._agent_state))
+            self._state_rpc.report_state(self._context, self._agent_state)
+            self._agent_state.pop('start_flag', None)
+        except AttributeError:
+            # This means the server does not support report_state
+            LOG.warn(_("Neutron server does not support state report."
+                       " Agent State reporting will be "
+                       "disabled."))
+            self._heartbeat.stop()
+            return
+        except Exception:
+            LOG.exception(_("Stopped reporting agent state!"))
+
+
+class RpcAgent(n_rpc.Service):
+
+    def __init__(
+            self, sc, host=None,
+            topic=None, manager=None, report_state=None):
+        super(RpcAgent, self).__init__(host=host, topic=topic, manager=manager)
+        if report_state:
+            self._report_state = AgentReportState(self.report_state)
+
+    def start(self):
+        LOG.debug(_("RPCAgent listening on %s" % (self.identify)))
+        super(RpcAgent, self).start()
+
+    def report_state(self):
+        if hasattr(self, '_report_state'):
+            LOG.debug(_("Agent (%s) reporting state" % (self.identify())))
+            self._report_state.report_state()
+
+    def identify(self):
+        return "(host=%s,topic=%s)" % (self.host, self.topic)
+
+
+class RpcAgents(object):
+
+    def __init__(self):
+        self.services = []
+        self.launchers = []
+
+    def add(self, agents):
+        for agent in agents:
+            LOG.debug(_("New RPC Agent %s" % (agent.identify())))
+        self.services.extend(agents)
+
+    def launch(self):
+        for s in self.services:
+            LOG.debug(_("Launching a rpc service %s" % (s.identify())))
+            l = os_service.launch(cfg.CONF, s)
+            self.launchers.extend([l])
+
+    def wait(self):
+        for l in self.launchers:
+            l.wait()
+
+    def report_state(self):
+        for agent in self.services:
+            agent.report_state()
+
+
```

Comment:
Will add comments.
Agent reporting is periodic reporting --- Actually some of the RPC Endpoints we have are neutron *aaS agents -- one requirement for *aas agent is to report its state periodically. In our design each agent could be a nsf-module or multiple agents can be in single module, so if it was allowed for each one to write their own logic it was little difficult. since this was common requirement for most of the code, state reporting is incorporated in the framework- to handle it in single common place to make it less error prone.
To report state like this, we need a periodic task -- hence another periodic task is started to handle report stating and there is one more periodic task which handles the event polling.